crates/libm-test/src/gen/domain.rs

```diff
@@ -0,0 +1,207 @@
+//! A generator that makes use of the `Domain` trait
+//!
+//! Does the following:
+//!
+//! - Check 100 values near each of the bounds (default)
+//! - If there are defined asymptotes, check those
+//! - Figure out a number of tests to do within the domain. If exhaustive, check all (but skip
+//!   NaNs?)
+//! - Check near zero if defined
+//! - If unbounded, ensure that real inputs do not produce any NaNs
+//! - If periodic, check that results are identical for a few periods (?)
+
+use std::iter;
+use std::ops::Bound;
+
+use libm::support::{Float, MinInt};
+
+use crate::domain::{Domain, HasDomain};
+use crate::{FloatExt, MathOp, logspace};
+
+/// Number of values near an interesting point to check.
+const AROUND: usize = 100;
+
+/// Number of tests to run.
+const NTESTS: usize = {
+    if cfg!(optimizations_enabled) {
+        if crate::emulated()
+            || !cfg!(target_pointer_width = "64")
+            || cfg!(all(target_arch = "x86_64", target_vendor = "apple"))
+        {
+            // Tests are pretty slow on non-64-bit targets, x86 MacOS, and targets that run
+            // in QEMU.
+            100_000
+        } else {
+            5_000_000
+        }
+    } else {
+        // Without optimizations just run a quick check
+        800
+    }
+};
+
+/// Some functions have infinite asymptotes, limit how many we check.
+const MAX_ASYMPTOTES: usize = 10;
```

Comment:
I need to refactor this all a bit considering I just copied and pasted NTESTS from the random tests. Still brainstorming how to do this (input appreciated) but I'm thinking maybe something like:

- Have `N` number of tests by default, which is 5M in the existing code (but probably needs to be reduced based on the below)
- `f32` unary functions run `N` random tests
- `f64` and `f128` unary functions run `N * 4` random tests
- If the function takes two inputs, multiply the number of tests by 4. If it takes three inputs, multiply by 8.
- If a domain test exists for the function, run `N` domain-based tests and reduce the number of random tests by a factor of 100
- If an exhaustive test for `f32` (not yet implemented) or high-iteration test (`f64`/`f128` or multi-input functions, also not implemented) should be run, still run the "interesting points" portion of domain-based tests but replace the `logspace` tests with whatever fits. I should probably split this into two generators rather than chaining...

Basically in all cases, either the exhaustive tests or the logspace are going to consume the bulk of time. But we still want to run "interesting points" and the random tests in hopes that they will find errors earlier than waiting for the whole exhaustive check to run. Also random tests should cover some signaling NaNs.

Reply:
I have this as part of https://github.com/rust-lang/libm/pull/364.